src/coreclr/System.Private.CoreLib/src/System/Threading/Interlocked.CoreCLR.cs

```diff
@@ -42,6 +42,24 @@ public static partial class Interlocked
         #endregion
 
         #region Exchange
+        /// <summary>Sets a 8-bit unsigned integer to a specified value and returns the original value, as an atomic operation.</summary>
+        /// <param name="location1">The variable to set to the specified value.</param>
+        /// <param name="value">The value to which the <paramref name="location1"/> parameter is set.</param>
+        /// <returns>The original value of <paramref name="location1"/>.</returns>
+        /// <exception cref="NullReferenceException">The address of location1 is a null pointer.</exception>
+        [Intrinsic]
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        public static extern byte Exchange(ref byte location1, byte value);
```

Comment:
Instead of creating an FCall here, I think we should consider the cost of a QCall and applying `SuppressGCTransition`. Unless there is a clear measurable benefit for an FCall we should be deferring to QCalls.

Reply:
> I feel like that could be done in a separate PR to not add any additional parts to this PR.

Sure. It is early enough in .NET 9, that that is a fair approach. Look forward to that PR.

Reply:
> Unless there is a clear measurable benefit for an FCall we should be deferring to QCalls.

As for this, the fallback path should only be used on RISC-V, ARM32 and WASM (with NativeAOT-LLVM) AFAIR. So none of those are currently of huge performance concern.

Reply:
Converting this method to QCall would leave a lot of perf on the table (the `ref` argument would require marshalling among other things).

It agre that the FCalls added in this PR are JIT intrinsics on x64 and arm64, but there are other pre-existing perf-critical System.Interlocked FCalls that are not JIT intrinsics and that won't be easy to turn into JIT intrinsics.

We should be considering all System.Interlocked FCalls - including the perf critical ones - for any conversions.

Reply:
> Converting this method to QCall would leave a lot of perf on the table (the ref argument would require marshalling among other things).

For this it would be a `fixed` statement and then call with a pointer. I agree there is cost but I think understanding how much is perhaps worth the investigation for something as minor as an interlocked op on a byte or short. Obviously not in this PR, but it does seem a bit hand wavey to say perf-critical without a known benchmark to follow up on.

Reply:
I agree that any changes in the pattern should be measured. I am pretty sure that the QCall overhead would be prohibitive for `Interlocked.CompareExchange(ref object, object, object)` and `Interlocked.CompareExchange(ref T, T, T)`.